### ISSUE TITLE
rustdoc reproducibility due to nondeterministic anon.llvm.<hash>

### ISSUE THREAD

**Issue by sundeep-kokkonda** [rust-lang/rust#126976]
I am facing a reproducibility issue with _rustdoc_ in Yocto project. The issue is happening due to a different _llvm.<**hash**>_ generation between the builds.

![anon-1](https://github.com/rust-lang/rust/assets/56015745/514181ff-35e8-4eef-96f8-3b566df2dc2f)
![anon-2](https://github.com/rust-lang/rust/assets/56015745/56ad7504-84dc-4911-8ef2-ac872a13aca4)

What could be the reason for this _llvm.<**hash**>_ is nondeterministic. Any optimization settings causing this? (I tried the PGO enabling/disabling but has no effect).

[Info] - This issue we've seen in past https://internals.rust-lang.org/t/rustdoc-binary-is-not-reproducible/20027. There we concluded the issue is fixed by _disabling PGO_ and setting _tools = ["rust-demangler",]_ but it is not fixed. Somehow with those settings _rustdoc_ is not generating at all. (We found that by removing _tools = ["rust-demangler",]_ the _rustdoc_ is generated)

**Comment by bjorn3** [comment]
The `.llvm.*` suffixes are generated by ThinLTO (which is enabled by default in the form of thin-local-lto in release mode). I don't know why it is non-deterministic though.

**Comment by sundeep-kokkonda** [comment]
Hello,

I did further analysis of this issue. Below are my observations,
- Issue is happening when the rust is built in different directory paths.
- Issue is reproduced on rust upstream sources as well (I made a small script to reproduce the issue w/ upstream sources on 1.75. Rename the attachment to 'rustdoc-repro.sh' and run '_sh rustdoc-repro.sh_'). It's reproducing on the 1.78 as well.
- The dump of `librustdoc-xxxx.rlib`'s shows that the `*.rcgu.o` are with different sizes. I tried the builds by disabling the rcgu's with `codegen-units = 1`. But still the compiled rustdoc with single rcgu also differed.
The objdump shows the _.rmeta_ has different size. But I couldn't disassemble and see what is in .rmeta.
- The increase of `codegen-units = 256` or `lto = off `or `opt-level = 0` also doesn't solve the problem.

**Can I get some help here that, how the rustdoc depends on build path and any pointers on the issue?**

rcgu size difference (build path name diff's are ignored so not visible in comparison)
![rcgu-sizes](https://github.com/rust-lang/rust/assets/56015745/cc4230a4-d3b7-4dfc-bfdf-52fc65727d9e)


[rustdoc-repro.txt](https://github.com/user-attachments/files/16156667/rustdoc-repro.txt)

**Comment by sundeep-kokkonda** [comment]
The objdump shows diff in `rmeta `sizes also. I suspect some metadata causing this issue. Is there anyway I can read/disassemble the contents of meta  file.

![rmeta-size-diff](https://github.com/rust-lang/rust/assets/56015745/a0f581e1-7c04-4288-8b49-9c48dab91c9d)

**Comment by sundeep-kokkonda** [comment]
@bjorn3 Any pointers here on this issue?

**Comment by bjorn3** [comment]
No clue unfortunately.

**Comment by sundeep-kokkonda** [comment]
> I have no clue what the problem is unfortunately.

May be your question is about a bit more detailed problem description?

The '_rustdoc_' binary is not reproducible when the rust sources are built in two different directories.  This is occurring with rust upstream sources (I verified on 1.78 sources).
I've a small script (with 1.75) attached in https://github.com/rust-lang/rust/issues/126976#issuecomment-2219718171 that will reproduce the issue. And, I've my observations also listed there.

In short summary, the _rcgu_'s generated between the builds are not identical on _rustdoc_ crate, which is not the case with other crates.

**Comment by bjorn3** [comment]
With problem I meant what the cause of this problematic behavior is. Should have been clearer.

**Comment by kanavin** [comment]
I also want to note that we currently work around the issue with this patch to rust's build system. e.g. setting -Clto=off for rustdoc only:

```
From ce68809d41291f671b440abce41f8f71c95428aa Mon Sep 17 00:00:00 2001
From: Alexander Kanavin <alex@linutronix.de>
Date: Mon, 3 Feb 2025 20:06:46 +0100
Subject: [PATCH] src/core/build_steps/tool.rs: switch off lto for rustdoc

For reasons currently unknown, librustdoc binary ends up with
non-reproducible .llvm.<number> suffixes in its symbols - but
not any other binary.

Disabling lto avoids creating these suffixes. More info about the option:
https://doc.rust-lang.org/rustc/codegen-options/index.html#lto

As seen below, there's a comment in the source tree saying not to tweak the options
but this only creates a mix of lto and non-lto optimized binary objects from
various crates, which should be safe to mix.

Upstream-Status: Inappropriate [reported at https://github.com/rust-lang/rust/issues/134589]
Signed-off-by: Alexander Kanavin <alex@linutronix.de>
---
 src/bootstrap/src/core/build_steps/tool.rs | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/src/bootstrap/src/core/build_steps/tool.rs b/src/bootstrap/src/core/build_steps/tool.rs
index 087df2f8a..00790affb 100644
--- a/src/bootstrap/src/core/build_steps/tool.rs
+++ b/src/bootstrap/src/core/build_steps/tool.rs
@@ -635,7 +635,7 @@ impl Step for Rustdoc {
         }
 
         // NOTE: Never modify the rustflags here, it breaks the build cache for other tools!
-        let cargo = prepare_tool_cargo(
+        let mut cargo = prepare_tool_cargo(
             builder,
             build_compiler,
             Mode::ToolRustc,
@@ -645,6 +645,7 @@ impl Step for Rustdoc {
             SourceType::InTree,
             features.as_slice(),
         );
+        cargo.rustflag("-Clto=off");
 
         let _guard = builder.msg_tool(
             Kind::Build,
-- 
2.39.5

```

**Comment by workingjubilee** [comment]
Revisiting consequent to [the initial #yocto discussion](https://rust-lang.zulipchat.com/#narrow/channel/557100-yocto/topic/2026-01-12.20meeting.20summary/with/568250963) on Zulip: the non-reproducible part leaking state into the LTO hashes appears to be build path, specifically.

This may need an LLVM fix.

**Comment by paradoxicalguy** [comment]
i can confirm this issue extends beyond rustdoc to `librustc_driver` under default bootstrap settings on x86_64-linux-gnu.
tested it with two stage2 builds in different workspace directories and no manual lto configuration, the anon prefix is identical but `.llvm.<hash>` suffixes differ consistently:
```
-anon.b9f2e6cb9cb63e5196a37045d61b280c.31.llvm.9551890413255510791
+anon.b9f2e6cb9cb63e5196a37045d61b280c.31.llvm.4464301409935140311
```
confirmed that bootstrap defaults to RustcLto::ThinLocal for the compiler in release mode, same as rustdoc. 
stripping debug info does not resolve the hash mismatch either, the gnu build id still differs between builds, the suffixes propogate into `.text` symbol references and GNU Build ID.

**Comment by windsunil** [comment]
We root-caused one instance of this: LLVM's LoopUnroll iterated a DenseMap, keyed by BasicBlock*, so phi operand order depended on the compiler's own memory layout. That changed 2 bytes of bitcode and flipped the ThinLTO module hash. rustc's pre-optimisation bitcode was byte-identical, so rustc wasn't at fault.

Fixed upstream by llvm/llvm-project@965f9d87, which is in LLVM 23 - so rust main has it. LLVM 22 and earlier don't. With it backported our reproducibility test passed 9 times running.
